### PR TITLE
add UT for storageclass

### PR DESCRIPTION
**Pull request by liyuerich**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

add UT for kubernetes/pkg/volume/util/storageclass, UT coverage can increase from 30.1% to 32.7%

#### Which issue(s) this PR fixes:

Fixes # NONE

#### Special notes for your reviewer:
NONE

#### Does this PR introduce a user-facing change?

```
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```
NONE
```

### CODE REVIEW

File: pkg/volume/util/storageclass_test.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	"fmt"
+	"testing"
+	"time"
+
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func TestGetDefaultClass(t *testing.T) {
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:30]: A better pattern to write test cases in go:

```go
func TestXXX(t *testing.T) {
  testCases := []struct {
    name string     // the name of a test case
    xxx XXX         // input args
    expectYYY YYY   // expect something
  }{
    {},
    {}
  }

  for _, testCase := range testCases {
    t.Run(testCase.name, func (t *testing.T){
      // do something
    })
  }
}
```

File: pkg/volume/util/storageclass_test.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			Annotations: map[string]string{
+				IsDefaultStorageClassAnnotation: "true",
+			},
+			CreationTimestamp: metav1.Time{Time: timestamp},
+		},
+	}
+}
+
+// MockLister implements the Lister interface for testing
+type MockLister struct {
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:98]: An example (mock lister for storageclass):

```go
informerFactory := informers.NewSharedInformerFactory(nil, controller.NoResyncPeriodFunc())
for _, c := range testCase.classes {
  informerFactory.Storage().V1().StorageClasses().Informer().GetStore().Add(c)
}
lister :=  informerFactory.Storage().V1().StorageClasses().Lister()
```

File: pkg/volume/util/storageclass_test.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+func TestGetDefaultClass(t *testing.T) {
+
+	var (
+		sc1 *storagev1.StorageClass = &storagev1.StorageClass{
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:33]:
```suggestion
		sc1 = &storagev1.StorageClass{
```

File: pkg/volume/util/storageclass_test.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "my-storage-class1",
+				Annotations: map[string]string{
+					"a": "b",
+				},
+			},
+		}
+		sc2 *storagev1.StorageClass = &storagev1.StorageClass{
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:41]: ditto

File: pkg/volume/util/storageclass_test.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "my-storage-class2",
+				Annotations: map[string]string{
+					"a": "b",
+				},
+			},
+		}
+
+		sc3 *storagev1.StorageClass = &storagev1.StorageClass{
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:50]: ditto

File: pkg/volume/util/storageclass_test.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "my-storage-class3",
+				Annotations: map[string]string{
+					IsDefaultStorageClassAnnotation: "true",
+				},
+				CreationTimestamp: metav1.Time{Time: time.Now().Add(-1 * time.Hour)},
+			},
+		}
+
+		sc4 *storagev1.StorageClass = &storagev1.StorageClass{
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:60]: ditto

File: pkg/volume/util/storageclass_test.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "my-storage-class4",
+				Annotations: map[string]string{
+					IsDefaultStorageClassAnnotation: "true",
+				},
+				CreationTimestamp: metav1.Time{Time: time.Now().Truncate(time.Hour)},
+			},
+		}
+
+		sc5 *storagev1.StorageClass = &storagev1.StorageClass{
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:70]: ditto

File: pkg/volume/util/storageclass_test.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+			classes: []*storagev1.StorageClass{sc1, sc2, sc3, sc4, sc5},
+			expect:  sc4,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			informerFactory := informers.NewSharedInformerFactory(nil, controller.NoResyncPeriodFunc())
+			for _, c := range tc.classes {
+				t.Logf("c %v", c)
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:114]: remove this message

File: pkg/volume/util/storageclass_test.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+				informerFactory.Storage().V1().StorageClasses().Informer().GetStore().Add(c)
+			}
+			lister := informerFactory.Storage().V1().StorageClasses().Lister()
+			actual, err := GetDefaultClass(lister)
+
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:119]: remove redundant empty line

File: pkg/volume/util/storageclass_test.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+		},
+
+		{
+			name:    "one default storage class",
+			classes: []*storagev1.StorageClass{sc1, sc2, sc3},
+			expect:  sc3,
+		},
+
+		{
+			name:    "three default storage class, newest frist,  class name second",
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:104]: split this to two. one for different creationTimestamp. another for different name with same creationTimestamp

File: pkg/volume/util/storageclass_test.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+			informerFactory := informers.NewSharedInformerFactory(nil, controller.NoResyncPeriodFunc())
+			for _, c := range tc.classes {
+				t.Logf("c %v", c)
+				informerFactory.Storage().V1().StorageClasses().Informer().GetStore().Add(c)
+			}
+			lister := informerFactory.Storage().V1().StorageClasses().Lister()
+			actual, err := GetDefaultClass(lister)
+
+			if err != nil {
+				t.Errorf("Expected no error got %v", err)
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:121]:
```suggestion
				t.Errorf("Expected no error, got %v", err)
```

File: pkg/volume/util/storageclass_test.go [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+		},
+
+		{
+			name:    "two default storage class, different create time, return  newest frist",
+			classes: []*storagev1.StorageClass{sc3, sc4},
+			expect:  sc4,
+		},
+
+		{
+			name:    "two default storage class, same cteate time, return ascending order frist",
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:110]:
```suggestion
			name:    "two default storage class with same creation timestamp",
```

File: pkg/volume/util/storageclass_test.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+		},
+
+		{
+			name:    "one default storage class",
+			classes: []*storagev1.StorageClass{sc1, sc2, sc3},
+			expect:  sc3,
+		},
+
+		{
+			name:    "two default storage class, different create time, return  newest frist",
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:104]:
```suggestion
			name:    "two default storage class with different creation timestamp",
```

File: pkg/volume/util/storageclass_test.go [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+			for _, c := range tc.classes {
+				informerFactory.Storage().V1().StorageClasses().Informer().GetStore().Add(c)
+			}
+			lister := informerFactory.Storage().V1().StorageClasses().Lister()
+			actual, err := GetDefaultClass(lister)
+			if err != nil {
+				t.Errorf("Expected no error, got %v", err)
+				return
+			}
+			if !reflect.DeepEqual(actual, tc.expect) {
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:128]:
```suggestion
			if tc.expect != actual {
```

**Review by carlory — CHANGES_REQUESTED** [review]
some suggesting changes, otherwise LGTM.

File: pkg/volume/util/storageclass_test.go [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+		sc5 = &storagev1.StorageClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "my-storage-class5",
+				Annotations: map[string]string{
+					IsDefaultStorageClassAnnotation: "true",
+				},
+				CreationTimestamp: metav1.Time{Time: time.Now().Truncate(time.Hour)},
+			},
+		}
+	)
```
Comment (carlory) [inline pkg/volume/util/storageclass_test.go:78]:
```suggestion
	var (
	        t1 = time.Now()
	        t2 = time.Now().Add(1 * time.Hour)

		sc1 = &storagev1.StorageClass{
			ObjectMeta: metav1.ObjectMeta{
				Name: "my-storage-class1",
				Annotations: map[string]string{
					"a": "b",
				},
			},
		}
		sc2 = &storagev1.StorageClass{
			ObjectMeta: metav1.ObjectMeta{
				Name: "my-storage-class2",
				Annotations: map[string]string{
					"a": "b",
				},
			},
		}

		sc3 = &storagev1.StorageClass{
			ObjectMeta: metav1.ObjectMeta{
				Name: "my-storage-class3",
				Annotations: map[string]string{
					IsDefaultStorageClassAnnotation: "true",
				},
				CreationTimestamp: metav1.Time{Time: t1},
			},
		}

		sc4 = &storagev1.StorageClass{
			ObjectMeta: metav1.ObjectMeta{
				Name: "my-storage-class4",
				Annotations: map[string]string{
					IsDefaultStorageClassAnnotation: "true",
				},
				CreationTimestamp: metav1.Time{Time: t2},
			},
		}

		sc5 = &storagev1.StorageClass{
			ObjectMeta: metav1.ObjectMeta{
				Name: "my-storage-class5",
				Annotations: map[string]string{
					IsDefaultStorageClassAnnotation: "true",
				},
				CreationTimestamp: metav1.Time{Time: t2},
			},
		}
	)
```

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by liyuerich** [comment]
thanks @carlory  and @thockin

**Pull request merged**